### PR TITLE
[Model] Expose Phi3v num_crops as a mm_processor_kwarg

### LINKED ISSUE #7861: [Usage]: set num_crops in LVLM
### Your current environment

```text
The output of `python collect_env.py`
```

### How would you like to use vllm

How to set up the num_crops for LVLMs? For example, when initializing the processor for Ph-3.5-vision-instruct, the hugginface code looks like the following: 

```python
processor = AutoProcessor.from_pretrained(model_id, 
  trust_remote_code=True, 
  num_crops=4
) 
```

But I didn't find a way to set num_crops in vllm.

I have check the [pull request #7710](https://github.com/vllm-project/vllm/pull/7710), but did't find the solution there.

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by alex-jw-brooks**
FIX https://github.com/vllm-project/vllm/issues/7861. This PR should be merged after https://github.com/vllm-project/vllm/pull/8657; it exposes `num_crops` as a processor_kwarg for phi3v models (see last commit) and adds a bunch of tests to ensure it's properly handled everywhere  in:

- The max token count
- The dummy data
- The input processor
- The default input mapper (which wraps the default HF image processor)

### Examples
1. Offline batch inference
```python
from vllm import LLM, SamplingParams
from vllm.assets.image import ImageAsset

question = "What is the content of this image?"
prompt = f"<|user|>\n<|image_1|>\n{question}<|end|>\n<|assistant|>\n"
image = ImageAsset("cherry_blossom").pil_image.convert("RGB")

llm = LLM(
    model="microsoft/Phi-3-vision-128k-instruct",
    trust_remote_code=True,
    max_num_seqs=5,
    mm_processor_kwargs={"num_crops": 4}
)

sampling_params = SamplingParams(temperature=0.2, max_tokens=64)

outputs = llm.generate(
    {
        "prompt": prompt,
        "multi_modal_data": {"image": image}
    }, 
    sampling_params=sampling_params
)

for o in outputs:
    generated_text = o.outputs[0].text
    print(generated_text)


```
Sample response: `The image captures a serene scene of a tall, white tower with a golden dome, standing majestically against a clear blue sky. The tower is partially obscured by a tree adorned with pink cherry blossoms, adding a touch of nature's beauty to the urban landscape.`
The example for offline inference for phi3v has also been updated to pass it in case users end up looking at it.

2. Through the server
```bash
python vllm/entrypoints/openai/api_server.py \
    --device cuda \
    --model microsoft/Phi-3-vision-128k-instruct \
    --tokenizer microsoft/Phi-3-vision-128k-instruct \
    --trust-remote-code \
    --api-key token-abc123 \
    --max_model_len 32000 \
    --disable-frontend-multiprocessing \
    --mm_processor_kwargs '{"num_crops": 4}' &
```

Client:
```python
from openai import OpenAI

client = OpenAI(base_url="http://localhost:8000/v1", api_key="token-abc123")

completion = client.chat.completions.create(
  model="microsoft/Phi-3-vision-128k-instruct",
  messages=[
    {
        "role": "user", "content": [
          {"type": "image_url", "image_url": {"url": "https://github.com/haotian-liu/LLaVA/blob/1a91fc274d7c35a9b50b3cb29c4247ae5837ce39/images/llava_v1_5_radar.jpg?raw=true"}},
          {"type": "text", "text": "Describe this image. "},
        ]
    }
  ]
)

print(completion.choices[0].message)
```
ChatCompletionMessage(content=" The radar chart displays performance metrics for four different models across various evaluation datasets. Each axis represents a dataset, with metrics such as 'BLEU-4', 'MME', 'PPE', 'SMOG', 'TED-LIUM', 'VLEN', 'Vocabulary', and 'Word Error Rate' ranging from 0 to 100. The chart includes the following models: 'BLIP-2', 'InstructBLIP', 'Qwen-VL-Chat', and 'LLaVA-1.5'. Each model has a distinct line and color on the chart, with their performance in each dataset marked along the corresponding axis. Data points are annotated with their values. The chart is titled 'Performance Metrics of Translation Models on NLP Datasets' and includes a legend for model identification.", refusal=None, role='assistant', function_call=None, tool_calls=[])

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by DarkLight1337** [comment]
Marking this as draft to make it clear that #8657 should be merged first.

File: examples/offline_inference_vision_language.py [hunk 1]
```diff
@@ -87,6 +87,7 @@ def run_phi3v(question, modality):
         model="microsoft/Phi-3-vision-128k-instruct",
         trust_remote_code=True,
         max_num_seqs=5,
+        processor_kwargs={"num_crops": 16},
```
Comment (DarkLight1337) [inline examples/offline_inference_vision_language.py:90]: Can you link to the HF repo explaining how to use `num_crops`?

Please update the multi-image input example as well.

**Comment by DarkLight1337** [comment]
Otherwise the PR looks good to me (the tests pass locally). You can mark the PR as ready once you have addressed the above comment.

**Comment by alex-jw-brooks** [comment]
Awesome, sounds good, thanks! Updated both examples with a short description about `num_crops` and the recommended values for single/multi-frame for single image/multi-image, respectively, at least based on the docs for `Phi-3.5-vision-instruct`, which is what we use in the multi-image example. 

`Phi-3-vision-128k-instruct`(what we use in the single image example)  doesn't explicitly say in the README, but uses the value recommended by `Phi-3.5-vision-instruct` for single frame in its config, so I assume it's the same

**Review by DarkLight1337 — APPROVED** [review]

**Pull request merged**